libkernelflinger/tpm2_security.c

```diff
@@ -107,6 +108,63 @@ static const attribute_matrix_t config_table[MAX_NV_NUMBER] = {
 	}
 };
 
+
+static EFI_STATUS tpm2_get_capability(
+		IN      TPM_CAP                   Capability,
+		IN      UINT32                    Property,
+		IN      UINT32                    PropertyCount,
+		OUT     TPMI_YES_NO               * MoreData,
+		OUT     TPMS_CAPABILITY_DATA      * CapabilityData
+		)
+{
+	EFI_STATUS ret;
+	UINT32 i;
+	TPML_TAGGED_TPM_PROPERTY *prop;
+
+	ret = Tpm2GetCapability(Capability, Property, PropertyCount, MoreData, CapabilityData);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Call Tpm2GetCapability failed");
+		return ret;
+	}
+	prop = &CapabilityData->data.tpmProperties;
+	debug(L"TPM2 capability: 0x%08x, data.tpmProperties.count: %d, more data: %d",
+			bswap_32(CapabilityData->capability), bswap_32(prop->count), *MoreData);
+	for (i = 0; i < bswap_32(prop->count); i++) {
+		debug(L"prop %d: property: 0x%08x, value: 0x%08x", i,
+				bswap_32(prop->tpmProperty[i].property),
+				bswap_32(prop->tpmProperty[i].value));
+	}
+
+	return ret;
+}
+
+static EFI_STATUS tpm2_get_cap_permanent(TPMA_PERMANENT *per)
+{
+	EFI_STATUS ret;
+	TPMI_YES_NO more_data;
+	TPMS_CAPABILITY_DATA cap_data;
+	UINT32 value;
+	TPML_TAGGED_TPM_PROPERTY *prop;
+
+	ret = tpm2_get_capability(TPM_CAP_TPM_PROPERTIES, TPM_PT_PERMANENT, 1, &more_data, &cap_data);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Get TPM cap permanent failed");
+		return ret;
+	}
+	prop = &cap_data.data.tpmProperties;
+	if (bswap_32(prop->count) <= 0) {
+		// Can't read the data
+		error(L"Get empty TPM capability data of TPM_PT_PERMANENT");
+		ret = EFI_NOT_FOUND;
+		return ret;
+	}
+
+	value = bswap_32(prop->tpmProperty[0].value);
```

Comment:
Can't you now replace with something like:
*per  = (TPMA_PERMANENT)bswap_32(prop->tpmProperty[0].value);
and suppress the value intermediate variable.

Reply:
Will cause compile error:
 error: used type 'TPMA_PERMANENT' where arithmetic or pointer type is required